Mathlib/GroupTheory/GroupAction/SubMulAction/Combination.lean

```diff
@@ -0,0 +1,319 @@
+    ((s : Finset α) : Set α) = s := rfl
+
+theorem mem_coe_iff {s : Nat.Combination α n} {a : α} : a ∈ (s : Finset α) ↔ a ∈ s := .rfl
+
+theorem eq_iff_subset : s = t ↔ (s : Finset α) ⊆ (t : Finset α) := by
+  rw [Finset.subset_iff_eq_of_card_le (t.prop.trans_le s.prop.ge), Subtype.ext_iff]
+
+theorem exists_mem_notMem (hn : 1 ≤ n) (hα : n < ENat.card α) {a b : α} (hab : a ≠ b) :
+    ∃ s : n.Combination α, a ∈ s ∧ b ∉ s := by
+  have ha' : n ≤ Set.encard {b}ᶜ := by
+    rw [← not_lt, ← ENat.add_lt_add_iff_left (k := 1) (ENat.coe_ne_top _), not_lt]
+    rwa [← Set.encard_singleton b, Set.encard_add_encard_compl,
+      Set.encard_singleton, Set.encard_univ, add_comm,
+      ENat.add_one_le_iff (ENat.coe_ne_top _)]
+  obtain ⟨s, has, has', hs⟩ :=
+    Set.exists_superset_subset_encard_eq (s := {a}) (by simp [Ne.symm hab]) (by simpa) ha'
+  simp only [Set.singleton_subset_iff, Set.subset_compl_singleton_iff] at has has'
+  letI : Set.Finite s := Set.finite_of_encard_eq_coe hs
+  use ⟨Set.Finite.toFinset this, by
+      rw [mem_iff, ←ENat.coe_inj, ← hs, this.encard_eq_coe_toFinset_card]⟩
+  simp [has, has']
+
+variable (α n) in
+/-- `Nat.Combination α n` as a `SubMulAction` of `Finset α`. -/
+@[to_additive /--`Nat.Combination α n` as a `SubAddAction` of `Finsetα`.-/]
+def subMulAction [DecidableEq α] : SubMulAction G (Finset α) where
+  carrier := n.Combination α
+  smul_mem' g s := (Finset.card_smul_finset g s).trans
+
+@[to_additive]
+instance [DecidableEq α] : MulAction G (n.Combination α) :=
+  (subMulAction G α n).mulAction
+
+variable {G}
+
+@[to_additive (attr := simp)]
+theorem coe_smul [DecidableEq α] {n : ℕ} {g : G} {s : n.Combination α} :
+    ((g • s : n.Combination α) : Finset α) = (g • s) :=
+  SubMulAction.val_smul (p := subMulAction G α n) g s
+
+@[to_additive]
+theorem smul_ne_iff_exists_mem_notMem_smul [DecidableEq α] {s t : n.Combination α} {g : G} :
+    g • s ≠ t ↔ ∃ a ∈ (s : Finset α), a ∉ g⁻¹ • (t : Finset α) := by
+  rw [← Finset.not_subset, not_iff_not, ← coe_smul, ← eq_iff_subset]
+  exact smul_eq_iff_eq_inv_smul g
+
+theorem addAction_faithful {G : Type*} [AddGroup G] {α : Type*} [AddAction G α] {n : ℕ}
+    [DecidableEq α] (hn : 1 ≤ n) (hα : n < ENat.card α)
+    {g : G} (hg : (AddAction.toPerm g : Equiv.Perm α) ≠ 1) :
+    ∃ s : n.Combination α, g +ᵥ s ≠ s := by
+  have : ∃ (a : α), (g +ᵥ a : α) ≠ a := by
+    by_contra! h
+    apply hg
+    ext a
+    simpa only using h a
+  obtain ⟨a, ha⟩ := this
+  obtain ⟨s, has, has'⟩ := exists_mem_notMem hn hα (Ne.symm ha)
+  use s
+  intro h
+  apply has'
+  rwa [← h, ← mem_coe_iff, coe_vadd, vadd_mem_vadd_finset_iff, mem_coe_iff]
+
+theorem faithfulVAdd {G : Type*} [AddGroup G] {α : Type*} [AddAction G α] {n : ℕ}
+    [DecidableEq α] (hn : 1 ≤ n) (hα : n < ENat.card α)
+    [FaithfulVAdd G α] :
+    FaithfulVAdd G (n.Combination α) := ⟨fun {g₁ g₂} h ↦ by
+  simp only [vadd_eq_iff_eq_neg_vadd, vadd_vadd] at h
+  rw [← neg_inj, ← sub_eq_zero, sub_neg_eq_add]
```

Comment:
We seem to be missing API for `FaithfulSMul` and `FaithfulVAdd`. Something like this:
```
@[to_additive]
theorem _root_.faithfulSMul_iff {α β : Type*} [Group α] [MulAction α β] :
    FaithfulSMul α β ↔ (∀ a : α, (∀ b : β, a • b = b) → a = 1) := by
  refine ⟨fun h a ha ↦ h.eq_of_smul_eq_smul ?_, fun h ↦ ⟨fun {a₁ a₂} ↦ ?_⟩⟩
  · simpa only [one_smul]
  · simpa only [mul_smul, inv_smul_eq_iff, inv_mul_eq_one, eq_comm (a := a₂)] using h (a₂⁻¹ * a₁)
```

Reply:
This gives a proof like
```
  rw [faithfulVAdd_iff] at h ⊢
  intro g hg
  contrapose! hg
  apply addAction_faithful hn hα
  contrapose! hg
  apply h
  simpa [Equiv.Perm.ext_iff] using hg
```